### ISSUE TITLE
Function to convert a CategoricalArray into an array of integer levels

### ISSUE THREAD

**Issue by juliohm** [JuliaData/CategoricalArrays.jl#228]
As discussed on slack, it would be nice to have a function that takes a categorical array like:

```Julia
using CategoricalArrays

a = categorical(["AB", missing, "CD"])
```

and converts it to the corresponding integer array with the levels:

```Julia
b = intify(a) # [1, missing, 2]
```

This function that I called here `intify` would also handle missing values as demonstrated above, they would just be maintained in the conversion.

**Comment by nalimilan** [comment]
FWIW Pandas calls this [`codes`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Categorical.codes.html#pandas.Categorical.codes). Might be a too generic term to claim, though. Not sure.

**Comment by ablaom** [comment]
In MLJ, we call this `int` (but missings not implemented yet). 

I agree it makes sense to have this in CA. Slight preference to `intify` over `codes`. But I like `int`. We already have `float` and not `floatify`. 

```julia
julia> categorical(["A", "B", "C"])
3-element CategoricalArray{String,1,UInt32}:
 "A"
 "B"
 "C"

julia> v = categorical(["A", "B", "C"])
3-element CategoricalArray{String,1,UInt32}:
 "A"
 "B"
 "C"

julia> levels!(v, ["C", "B", "A"])
3-element CategoricalArray{String,1,UInt32}:
 "A"
 "B"
 "C"

julia> int(v)
3-element Array{UInt32,1}:
 0x00000003
 0x00000002
 0x00000001

julia> int(v, type=Int)
3-element Array{Int64,1}:
 3
 2
 1
```

**Comment by juliohm** [comment]
@ablaom it seems like the dev branch of ScientificTypes.jl has a function called `coerce`. Copying/pasting a snippet of code provided by @tlienart on slack:

```julia
julia> using CategoricalArrays
julia> a = categorical(["AB", missing, "CD"])
3-element CategoricalArray{Union{Missing, String},1,UInt32}:
 "AB"
 missing
 "CD"
julia> using ScientificTypes
julia> coerce(a, Count, verbosity=0)
3-element Array{Union{Missing, UInt32},1}:
 0x00000001
           missing
 0x00000002
```

**Comment by ablaom** [comment]
Yes, you can also do it that way (thanks to recent addition of @tlienart ).

**Comment by nalimilan** [comment]
I don't really like names mentioning "int" as we absolutely need to avoid the confusion with `convert(Vector{Int}, x)`, which works for e.g. `CategoricalArray{Int}` but with a completely different meaning. Also note that `float` works and is equivalent to calling `convert`.

**Comment by tlienart** [comment]
What  do you think of the way to do  it suggested in ScientificTypes i.e.:

```julia
julia> coerce(a, Count)
```

This effectively does an **order-respecting** conversion of `a`, also the other thing is that it carries the intent reasonably well, i.e. that we're re-interpreting the data (which is what ScientificTypes is all about).

**Comment by nalimilan** [comment]
Well, that's sounds cool in the context of your more general framework. But for CategoricalArrays in isolation I'd say we need something simpler, i.e. a single-purpose function.

**Comment by juliohm** [comment]
Any update regarding this issue? That is a very useful function to have.

**Comment by nalimilan** [comment]
We just have to find a good name. Maybe `levelcodes`?

**Comment by bkamins** [comment]
Following @nalimilan suggestion on Slack I would define `levelcode` or even `levelindex` (as it would be essentially an index in the return value from `levels` function) on `CategoricalValue` and then just broadcast it if needed for a whole vector.

**Comment by nalimilan** [comment]
See #239. I still somewhat like `levelcode`, which may be closer to the common terminology for this kind of thing. Do others have opinions?

**Comment by juliohm** [comment]
+1 for `levelcode`

**Comment by juliohm** [comment]
Any solution for this yet? I always come back to this issue in user code. It would be nice to have a temporary solution at least.

**Comment by juliohm** [comment]
As discussed on slack, it may be even appropriate to rename the function `levels` to `categories`, and then free the name `level` for the corresponding integer for the categories. So `categories` would be a list of possible categories, and if the user has the implicit assumption in his use case that these categories can be ordered, then `level` would return the index in the `categories` output.

**Comment by nalimilan** [comment]
As I noted on Slack, I'm absolutely opposed to changing `levels` to do something completely different from what it does now and in R. As for renaming `levels` to `categories`, maybe, but I'm hesitant since it will break lots of code and differ from R, for a limited gain (IMO).

Unfortunately we can't merge this before settling the naming discussion...

**Comment by juliohm** [comment]
As I understand we could have `categories` as the new name for `levels` and then choose a better name for the `levels` in my example above right? Maybe `index` or `int`?

**Comment by bkamins** [comment]
I would propose to just use `levelcode` (it seems at least @nalimilan + @juliohm prefer it over `levelindex`) and fix this issue with #239 after an update.

I feel that at this maturity of CategoricalArrays.jl significantly changing naming of major functions. In particular `levels` lives in DataAPI.jl so the change would be more global than only this package.

**Comment by juliohm** [comment]
Wouldnt this name change only cause a deprecation warning from levels to
categories ? And then the more natural name level is left for the ordering ?

**Comment by bkamins** [comment]
My feeling is that most people (in particular probably many R users) expect "level" to be a value not a number of this value in the order of possible values.

I think that even if `categories` is a better name than `levels` it is only minimally better, and as @nalimilan noted:
1. this change would impact a lot of existing code
2. this change would impact a lot of existing manuals
3. users from R expect `levels` to work as it works now
4. and (what is not mentioned there) even if it is not an R user I do not think that for most people when they would see the name of the function `level` then they naturally think that it returns "order of the value in the list of all possible values"

**Comment by juliohm** [comment]
I think we had this discussion on slack and the conclusions were the
following:

The name levels is jargon from the R community. Most papers in scientific
journals that we read in ML stats etc use the term categories to refer to
the unordered list of names. The proposal with "categories" in place of
"levels" seems adequate and less jargon from R.

The name levels as proposed above in the snippet of code is indeed
problematic because some users expect different semantics. To solve that I
suggested to instead use something like index or int.

So if you see my snippet of code above just replace levels with int or
index and we would just introduce one dep warning. I don't think many
packages would have difficulties updating.

Is this proposal ok?

**Comment by bkamins** [comment]
I feel that `int` is not good as @nalimilan commented earlier, `index` seems OK, but it feels a bit too common name (e.g. in DataFrames.jl `index` is used as a name to represent a completely different function).

I cannot say what is most common in ML. My feeling that "level" and "category" are both commonly used. Reading [Wikipedia](https://en.wikipedia.org/wiki/Categorical_variable) I find:

> Commonly (though not in this article), each of the possible values of a categorical variable is referred to as a level

(so it just states that there are two commonly used names for the same thing)

If we would agree to this definition then `levels` as collection of "level" values seems natural.

My point is that we have a decision here to change a status quo (from `levels` to `categories`). There are costs associated with this change (as listed above: need to change packages, deprecations in manuals/books etc.). Therefore it would be good to understand the benefits, like - do you believe that using `categories` instead of `levels` will lead to less confusion or easier user adoption in the future that justifies changing it.

(it would be a different discussion if we were to introduce this feature and were deciding what name is best to use, but as I have said we hare in "status quo" situation here which slightly changes things in my opinion)

**Comment by nalimilan** [comment]
It's true that "categories" seems to be used more often than "levels" in the first pages returned by a simple Web search, and that's the term used by Pandas and MATLAB. But the number of Google results for `"categorical variable" "levels"` is higher than for `"categorical variable" "categories"`, so...

In general, terminology varies so much between statistical fields that it's hard for any individual to know what's the dominant convention (if that concept even makes sense...).

Anyway, we could add a function now and discuss the broader renaming issue separately. It's going to be so disruptive if we change this that renaming a newly-introduced function will be a minor point.

**Comment by juliohm** [comment]
I agree that there is some urgency for merging this feature regardless of the name. Based on the Google search results, the term `categories` is more common for all kinds of scientists that are not statisticians, particularly statisticians that are familiar with R. There is an argument in favour of users coming from Pandas and MATLAB, which I think is a larger user base than R's?

My vote goes for `categories` + `index`. I understood that the alternative is `levels` + `levelcode`?

**Comment by nalimilan** [comment]
> I agree that there is some urgency for merging this feature regardless of the name. Based on the Google search results, the term `categories` is more common for all kinds of scientists that are not statisticians, particularly statisticians that are familiar with R. There is an argument in favour of users coming from Pandas and MATLAB, which I think is a larger user base than R's?

How do you see that? As I said, "categories" gives *fewer* results than "levels".

`index` is already used for something else in CategoricalArrays.

**Comment by juliohm** [comment]
>It's true that "categories" seems to be used more often than "levels" in
the first pages returned by a simple Web search, and that's the term used
by Pandas and MATLAB. But the number of Google results for "categorical
variable" "levels" is higher than for "categorical variable" "categories",
so...

That is not saying that the term categories is more often used?

**Comment by nalimilan** [comment]
Actually, exactly the reverse...

**Comment by juliohm** [comment]
I misunderstood the comment then. I had understood that categories was more
often used in general when the search is not restricted to categorical
variables.

**Comment by bkamins** [comment]
Anyway - given that the discussion is not simple - and that `index` is not really the name that can be flexibly used (it is common, it is used in CategoricalArrays.jl for a different purpose, it is also used in DataFrames.jl for different purposes) I propose the following:
* merge https://github.com/JuliaData/CategoricalArrays.jl/pull/239 as is - to have the functionality in place
* we can keep the "renaming" discussion here - if we come to the conclusion at some point that it is worth to rename functions it should be an easy deprecation process

**Comment by ablaom** [comment]
`levelcode`?

**Comment by bkamins** [comment]
Yes - `levelcode` is the current consensus when we keep `levels`.

**Comment by juliohm** [comment]
Out of curiosity, does R use the term `levels` for all kinds of categorical variables, even the nominal ones? To me the word `levels` has some order associated to it.

**Comment by bkamins** [comment]
```
> x <- factor(1:10)
> str(x)
 Factor w/ 10 levels "1","2","3","4",..: 1 2 3 4 5 6 7 8 9 10
> levels(x)
 [1] "1"  "2"  "3"  "4"  "5"  "6"  "7"  "8"  "9"  "10"
```

**Comment by juliohm** [comment]
Yes, these function names in R are really weird. I feel that people from Python, MATLAB, etc will need to lookup `factor` and `levels`. Maybe we could merge the `levels` and `levelcode` to get things moving knowing it is a sort of jargon from R.

**Comment by bkamins** [comment]
> Maybe we could merge the `levels` and `levelcode` to get things moving knowing it is a sort of jargon from R.

I agree - let us just merge it. And - then we can keep discussing the best naming separately, as changing the names would only mean a simple deprecation.

**Comment by pgagarinov** [comment]
@nalimilan @bkamins `levelcode` is missing in https://categoricalarrays.juliadata.org/latest/functionindex.html which makes it difficult to find.

**Comment by bkamins** [comment]
I have opened an issue to track this.
